### PR TITLE
Add growatt modbus sensor

### PR DESCRIPTION
**Pull request by leeuwte**
# What does this implement/fix? 

Adds the growatt modbus sensor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1732

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
# Example configuration entry
    uart:
      - id: uart1
        baud_rate: 9600
        tx_pin: GPIO1
        rx_pin: GPIO3

    modbus:
      uart_id: uart1
      flow_control_pin: GPIO4

    sensor:
      - platform: growatt_solar
        update_interval: 10s
        
        inverter_status:
          name: "Growatt Status Code"

        phase_a:
          voltage:
              name: "Growatt Voltage Phase A"
          current:
              name: "Growatt Current Phase A"
          active_power:
              name: "Growatt Power Phase A"
        phase_b:
          voltage:
              name: "Growatt Voltage Phase B"
          current:
              name: "Growatt Current Phase B"
          active_power:
              name: "Growatt Power Phase B"
        phase_c:
          voltage:
              name: "Growatt Voltage Phase C"
          current:
              name: "Growatt Current Phase C"
          active_power:
              name: "Growatt Power Phase C"

        pv1:
          voltage:
              name: "Growatt PV1 Voltage"
          current:
              name: "Growatt PV1 Current"
          active_power:
              name: "Growatt PV1 Active Power"

        pv2:
          voltage:
              name: "Growatt PV2 Voltage"
          current:
              name: "Growatt PV2 Current"
          active_power:
              name: "Growatt PV2 Active Power"

        active_power:
          name: "Growatt Grid Active Power"

        pv_active_power:
          name: "Growatt PV Active Power"

        frequency:
          name: "Growatt Frequency"

        energy_production_day:
          name: "Growatt Today's Generation"
          
        total_energy_production:
          name: "Growatt Total Energy Production"

        inverter_module_temp:
          name: "Growatt Inverter Module Temp"

```yaml
# Example config.yaml


```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/growatt_solar/growatt_solar.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+static const uint8_t MODBUS_CMD_READ_IN_REGISTERS = 0x04;
+static const uint8_t MODBUS_REGISTER_COUNT = 33;
+
+void GrowattSolar::update() { this->send(MODBUS_CMD_READ_IN_REGISTERS, 0, MODBUS_REGISTER_COUNT); }
+
+void GrowattSolar::on_modbus_data(const std::vector<uint8_t> &data) {
+  auto publish_1_reg_sensor_state = [&](sensor::Sensor *sensor, size_t i, float unit) -> void {
+    float value = encode_uint16(data[i * 2], data[i * 2 + 1]) * unit;
+    if (sensor != nullptr)
+      sensor->publish_state(value);
```
Comment (jesserockz) [inline esphome/components/growatt_solar/growatt_solar.cpp]: This way it doesnt need to encode even if the sensor is not set. (Same for the next one too)
```suggestion
    if (sensor == nullptr)
      return;
    float value = encode_uint16(data[i * 2], data[i * 2 + 1]) * unit;
    sensor->publish_state(value);
```

File: esphome/components/growatt_solar/growatt_solar.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#include "esphome/components/modbus/modbus.h"
+
+namespace esphome {
+namespace growatt_solar {
+
+static const float TWO_DEC_UNIT = 0.01;
+static const float ONE_DEC_UNIT = 0.1;
+
+class GrowattSolar : public PollingComponent, public modbus::ModbusDevice {
+ protected:
```
Comment (jesserockz) [inline esphome/components/growatt_solar/growatt_solar.h]: For consistency, can you move the `protected` block under the `public` one

File: esphome/components/growatt_solar/growatt_solar.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  struct GrowattPhase {
+    bool setup{false};
```
Comment (jesserockz) [inline esphome/components/growatt_solar/growatt_solar.h]: Please remove as you don't use the `setup` bools

File: esphome/components/growatt_solar/growatt_solar.h [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    sensor::Sensor *voltage_sensor_{nullptr};
+    sensor::Sensor *current_sensor_{nullptr};
+    sensor::Sensor *active_power_sensor_{nullptr};
+  } phases_[3];
+  struct GrowattPV {
+    bool setup{false};
```
Comment (jesserockz) [inline esphome/components/growatt_solar/growatt_solar.h]: Same here

**Review by jesserockz — CHANGES_REQUESTED** [review]
Mostly looks good, just a few minor things

File: esphome/components/growatt_solar/growatt_solar.cpp (hunk 1 above)
Comment (leeuwte) [reply esphome/components/growatt_solar/growatt_solar.cpp]: done!

File: esphome/components/growatt_solar/growatt_solar.h (hunk 2 above)
Comment (leeuwte) [reply esphome/components/growatt_solar/growatt_solar.h]: ok :D

File: esphome/components/growatt_solar/growatt_solar.h (hunk 3 above)
Comment (leeuwte) [reply esphome/components/growatt_solar/growatt_solar.h]: of course (copy paste 🤦 )

File: esphome/components/growatt_solar/growatt_solar.h (hunk 4 above)
Comment (leeuwte) [reply esphome/components/growatt_solar/growatt_solar.h]: of course (copy paste 🤦 ) 2

**Review by jesserockz — APPROVED** [review]

**Pull request merged**